server/src/main/java/org/opensearch/index/ShardIndexingPressureMemoryManager.java

```diff
@@ -0,0 +1,456 @@
+ * is greater than primary_parameter.node.soft_limit, then MemoryManager will evaluate deeper parameters for shards to identify any
+ * issues, such as throughput degradation (Secondary Parameter - 1) and time since last request was successful (Secondary Parameter - 2).
+ * This helps identify detect any duress state with the shard, requesting more memory.
+ *
+ * Secondary Parameters covered above:
+ * 1. ThroughputDegradationLimitsBreached - When the moving window throughput average has increased by a factor compared to
+ * the historical throughput average. If the factor by which it has increased is greater than the degradation limit threshold, this
+ * parameter is considered to be breached.
+ * 2. LastSuccessfulRequestDurationLimitsBreached - When the time since the last successful request completed is greater than the max
+ * timeout threshold value, while there a number of outstanding requests greater than the max outstanding requests then this parameter
+ * is considered to be breached.
+ *
+ * MemoryManager attempts to increase of decrease the shard limits in case the shard utilization goes below operating_factor.lower or
+ * goes above operating_factor.upper of current shard limits. MemoryManager attempts to update the new shard limit such that the new value
+ * remains withing the operating_factor.optimal range of current shard utilization.
+ *
+ */
+public class ShardIndexingPressureMemoryManager {
+    private final Logger logger = LogManager.getLogger(getClass());
+
+    /**
+     * Shard operating factor can be evaluated using currentShardBytes/shardLimits. Outcome of this expression is categorized as
+     * lower, optimal and upper boundary, and appropriate action is taken once the below defined threshold values are breached.
+     */
+    public static final Setting<Double> LOWER_OPERATING_FACTOR =
+        Setting.doubleSetting("shard_indexing_pressure.operating_factor.lower", 0.75d, 0.0d,
+            Setting.Property.NodeScope, Setting.Property.Dynamic);
+    public static final Setting<Double> OPTIMAL_OPERATING_FACTOR =
+        Setting.doubleSetting("shard_indexing_pressure.operating_factor.optimal", 0.85d, 0.0d,
+            Setting.Property.NodeScope, Setting.Property.Dynamic);
+    public static final Setting<Double> UPPER_OPERATING_FACTOR =
+        Setting.doubleSetting("shard_indexing_pressure.operating_factor.upper", 0.95d, 0.0d,
+            Setting.Property.NodeScope, Setting.Property.Dynamic);
+
+    /**
+     * This determines the max time elapsed since any request was processed successfully. Appropriate action is taken
+     * once the below below defined threshold value is breached.
+     */
+    public static final Setting<Integer> SUCCESSFUL_REQUEST_ELAPSED_TIMEOUT =
+        Setting.intSetting("shard_indexing_pressure.secondary_parameter.successful_request.elapsed_timeout", 300000,
+            Setting.Property.NodeScope, Setting.Property.Dynamic);
```

Comment:
Use `Setting<TimeValue>` instead, the current timeout doesn't capture units

Reply:
Its better if you name `secondary_parameter ` else it gets hard for users to understand and tune

Reply:
++